### PR TITLE
[Bug fix] [BEVFormer] Fix SCA double output projection, use deterministic lidar2img in tests

### LINKED ISSUE #55187: [BEVFormer] Fix SCA output projection and replace random lidar2img test geometry
## Context

Spatial Cross Attention (SCA) wraps a nested Multi-Scale Deformable Attention (MSDA) module. Both modules own a separate `output_proj` and the reference implementation applies the MSDA projection first, followed by the SCA projection.

Encoder tests currently use random `lidar2img` matrices. Arbitrary 4x4 matrices do not describe a physically valid camera and produce unrealistic sampling patterns: most valid sampling coordinates fall near the image border, where zero padding can hide an incorrect projection. A deterministic camera rig moves most samples into the interpolated image region and exposes encoder PCC failures.

Replacing random `lidar2img` generation is therefore a prerequisite for trusting any encoder PCC result, not a convenience for this fix alone. Every downstream test that depends on projection geometry inherits the same requirement.

## Problem

- [SCA parameter preprocessing stores the outer](https://github.com/tenstorrent/tt-metal/blob/main/models/experimental/bevformer/tt/model_preprocessing.py#L295-L318) `output_proj` [at the top level, flattens nested MSDA parameters into the same namespace, and skips the nested](https://github.com/tenstorrent/tt-metal/blob/main/models/experimental/bevformer/tt/model_preprocessing.py#L295-L318) `output_proj`.
- [SCA passes that flat parameter namespace to the inner MSDA module](https://github.com/tenstorrent/tt-metal/blob/main/models/experimental/bevformer/tt/tt_spatial_cross_attention.py#L83).
- [MSDA applies](https://github.com/tenstorrent/tt-metal/blob/main/models/experimental/bevformer/tt/tt_ms_deformable_attention.py#L328-L330) `self.params.output_proj`, which resolves to the SCA projection, and [SCA applies the same projection again](https://github.com/tenstorrent/tt-metal/blob/main/models/experimental/bevformer/tt/tt_spatial_cross_attention.py#L329-L331).
- The TT path therefore applies `SCA.output_proj` twice and never applies `MSDA.output_proj`, unlike the [reference MSDA](https://github.com/tenstorrent/tt-metal/blob/main/models/experimental/bevformer/reference/ms_deformable_attention.py#L258) and [reference SCA](https://github.com/tenstorrent/tt-metal/blob/main/models/experimental/bevformer/reference/spatial_cross_attention.py#L175-L204).

## Proposed direction

- Replace random `lidar2img` [test inputs](https://github.com/tenstorrent/tt-metal/blob/main/models/experimental/bevformer/tests/pcc/test_encoder.py#L46-L69) with deterministic camera-rig geometry that exercises real interpolation. Remove the random-matrix path rather than keeping it as an alternative; leaving it available lets unrealistic sampling patterns hide correctness problems in any test that reuses it.
- Store all nested MSDA parameters, including its `output_proj`, under `deformable_attention`.
- Preserve the nested namespace when converting both freshly preprocessed and cached parameters to objects.
- Pass only `params.deformable_attention` to the inner MSDA module.
- Keep existing PCC thresholds; lowering them would make the incorrect projection part of the expected behavior.

## Acceptance criteria

- [x] Nested MSDA and outer SCA `output_proj` weights are both loaded and remain independently addressable.
- [x] The inner MSDA applies its own projection, followed by the outer SCA projection in the same order as the reference.
- [x] Deterministic camera-rig encoder tests pass without lowering thresholds: `nuscenes_tiny` PCC >= 0.996, `carla_tiny` PCC >= 0.995, and base configurations at their existing thresholds.
- [x] Standalone SCA, MSDA, point-sampling, temporal-attention, layer, and encoder PCC tests pass at their configured thresholds.
- [x] Fresh and cached parameter-preprocessing paths produce the same nested parameter structure.
- [x] No existing PCC threshold is reduced or re-baselined.
- [x] No encoder test constructs `lidar2img` from arbitrary random 4x4 matrices; every projection matrix comes from a valid camera rig.

## References

- [Deterministic camera-rig prototype](https://github.com/tenstorrent/tt-metal/commit/752990330b96edf4e0ca0c2125592d932f1998cb)
- [Parameter-plumbing fix prototype](https://github.com/tenstorrent/tt-metal/commit/d897012f5cc8b276f6afa6f55ad22818d65764c8)
- [Layer, encoder, and diagnostic test prototype](https://github.com/tenstorrent/tt-metal/commit/a213d3c1135bc66baff97bc57f2712e863c83ef6)
- [Encoder PCC tests](https://github.com/tenstorrent/tt-metal/blob/main/models/experimental/bevformer/tests/pcc/test_encoder.py)
- [SCA PCC tests](https://github.com/tenstorrent/tt-metal/blob/main/models/experimental/bevformer/tests/pcc/test_spatial_cross_attention.py)
- [MSDA PCC tests](https://github.com/tenstorrent/tt-metal/blob/main/models/experimental/bevformer/tests/pcc/test_ms_deformable_attention.py)

### PR DESCRIPTION
**Pull request by ctr-mmicicTT**
### PR Category
Bug fix

### Summary
Closes #55187 (parent: #55048)

Two coupled problems in the BEVFormer encoder:

1. `SpatialCrossAttention` and its nested deformable attention each own an
   `output_proj`. Preprocessing flattened both into one namespace, so the nested
   module's weights were dropped and it projected with the SCA matrix, which SCA
   then applied a second time. The reference applies MSDA's projection first,
   then SCA's.
2. Encoder tests built `lidar2img` from `torch.randn(4, 4)`. Arbitrary matrices
   describe no physical camera: most samples land on the zero-padded image
   border, and `max_len` varied 1268-4563 across runs.

The second hid the first — only with sampling off the borders does the wrong
projection cost enough PCC to fail a threshold.

Fix: keep nested MSDA parameters (including its `output_proj`) under
`deformable_attention` on both the fresh and cached preprocessing paths, and
pass only that namespace to the inner module. Replace random `lidar2img` with a
deterministic nuScenes-like camera rig (synthetic 360-degree setup for datasets
without one), intrinsics scaled to the configured input resolution.

Encoder PCC: `nuscenes_tiny` 0.994607 -> 0.999298, `carla_tiny` 0.994937 ->
0.999444. Both were failing. No threshold lowered or re-baselined. PCC tests run
on this branch and pass.

### Notes for reviewers
- The random-`lidar2img` path is deleted, not kept behind a flag. Leaving it
  available lets unrealistic sampling hide correctness bugs in any test reusing
  it.
- The rig consumes no RNG state, so `bev_query`, `bev_pos` and
  `camera_features` are bit-identical between runs; only geometry moves.
- `max_len` is now a stable 2472 (was 3660). Perf captures across that boundary
  are not comparable.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:ctr-mmicic/55187-fix-sca-output-projection)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:ctr-mmicic/55187-fix-sca-output-projection)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:ctr-mmicic/55187-fix-sca-output-projection)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:ctr-mmicic/55187-fix-sca-output-projection)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:ctr-mmicic/55187-fix-sca-output-projection)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:ctr-mmicic/55187-fix-sca-output-projection)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:ctr-mmicic/55187-fix-sca-output-projection)

### CODE REVIEW

File: models/experimental/bevformer/config/encoder_config/camera_rig.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+        Tensor of shape ``[len(specs), 4, 4]``.
+    """
+    return torch.stack([_intrinsic_matrix(spec, input_size, dtype) @ _extrinsic_matrix(spec, dtype) for spec in specs])
+
+
+def camera_rig_for_dataset(dataset_config) -> Tuple[CameraSpec, ...]:
+    """Pick the rig recorded for a dataset, falling back to a synthetic ring."""
+    if dataset_config.name.startswith("nuscenes") and dataset_config.num_cams == len(NUSCENES_CAMERA_RIG):
+        return NUSCENES_CAMERA_RIG
+    return ring_camera_rig(dataset_config.num_cams, dataset_config.input_size)
```
[the remark this answers is not in the document]
Comment (milos-micic) [reply models/experimental/bevformer/config/encoder_config/camera_rig.py:201]: Fixed in https://github.com/tenstorrent/tt-metal/commit/2cffb1fe244fb67977d1b109a88e3ed1ea6ec06a

File: models/experimental/bevformer/tt/model_preprocessing.py [hunk 2]
```diff
@@ -288,7 +304,7 @@ def preprocess_spatial_cross_attention_parameters(
     cached_params = _manage_cache_load(cache_file_name, device)
     if cached_params is not None:
         # Convert cached params to object format for dot notation access
-        return convert_parameterdict_to_object(cached_params)
+        return _convert_sca_parameters_to_object(cached_params)
```
[the remark this answers is not in the document]
Comment (milos-micic) [reply models/experimental/bevformer/tt/model_preprocessing.py:307]: Cache is not used anywhere at the moment so versioning of cached data structure is not needed yet.

**Review by mbezuljTT — APPROVED** [review]
